src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.Tracing;
+
+namespace System.Net.Sockets
+{
+    [EventSource(Name = "System.Net.Sockets")]
+    internal sealed class SocketsTelemetry : EventSource
+    {
+        public static readonly SocketsTelemetry Log = new SocketsTelemetry();
+
+        [NonEvent]
+        public static bool IsEnabled(EventLevel level)
+        {
+            return Log.IsEnabled(level, EventKeywords.All);
+        }
+
+        [NonEvent]
+        public void ConnectStart(Internals.SocketAddress address)
+        {
+            ConnectStart(address.ToString());
+        }
+
+        [NonEvent]
+        public void ConnectStart(EndPoint address)
+        {
+            ConnectStart(address.ToString());
+        }
+
+        [Event(1, Level = EventLevel.Informational)]
+        public void ConnectStart(string? address)
+        {
+            WriteEvent(eventId: 1, address ?? "");
+        }
+
+        [Event(2, Level = EventLevel.Informational)]
+        public void ConnectStop()
+        {
+            ConnectStopInternal();
+        }
+
+        [Event(3, Level = EventLevel.Error)]
+        public void ConnectFailed()
```

Comment:
There is `SocketError` or `Exception`, but we don't log the details here because the goal is to provide telemetry for operation tracing/timings not debugging. Please, see how it has been already done in [HttpTelemetry](https://github.com/davidsh/runtime/blob/e5274e41fe13a8cfeeb4e87e39a1452cf8968f8c/src/libraries/System.Net.Http/src/System/Net/Http/HttpTelemetry.cs#L51).

Reply:
If the goal is solely to do timing analysis and not facilitate failure analysis then do we need a connect failed event at all? I assumed the purpose of this event having a different level is so that we can do failure analysis even while not doing timing analysis.

I probably should have asked the same question about the HttpTelemetry, I missed it there and caught it here : )

Reply:
It's not only about timing, but also about number or rate of Successes/Failures, this is why we need `Failed`.

Reply:
Exactly. This is my same question/concern with the design. I don't understand in what circumstance I care to differentiate why a connect stopped but not have more details on the failures.

Reply:
Knowing the failure rate is quite useful for live service health monitoring/incident analysis because it's not uncommon for health to start degrading gradually so by observing failure rate you can catch it.
Having that said, I'm not against adding failure details (it's also useful), but we will need also to update implemented code (like HttpTelemetry).

Reply:
Failure rate is fine, but it should be part of the stop event, no? If we then add a second event for failure, it should include details.

Reply:
[As discussed above](https://github.com/dotnet/runtime/pull/38620#discussion_r453504085), having a separate Failed event makes sense in this case taking into account that Stop/Fail have different levels.

Reply:
SocketError and Exception (if any) are logged now.

Reply:
The only reason I see for Failure() to have a different level is if it plays a role in failure analysis. I recommend that is what we should do (for http as well), but if we didn't define the scenarios that way then the dominos would start falling:
Failure has no reason to be error -> Failure moves to be Info level -> Failure has no reason to be distinct event from stop -> failure event is removed and stop gains a pass/fail flag

Reply:
Failure **is** needed for failure analysis here and in HttpTelemetry as well.

Reply:
I suspect we may be using the terminology differently and I'm happy to explain further if it would be helpful, but after you added the Exception/status code to the ConnectFailed() event my concerns on this are resolved. Thanks!

Reply:
Let's say I enable start/stop events and the failed event.  How do I correlate an exact failed with an exact start/stop?  What data is included in all of these events that makes that feasible?

Reply:
Yes, EventSource infra automatically creates an activity for each start/stop pair and all nested events gets linked to parent ones (multi-level nesting is supported). Each activity has ID that is logged and used for correlation.

Reply:
If you want to see an example of the Activity tracing in action, there is sample code and some output included as the repro of this bug I filed last night while playing with it more: https://github.com/dotnet/runtime/issues/39353